packages/roosterjs-editor-dom/lib/edit/pasteTable.ts

```diff
@@ -0,0 +1,74 @@
+import findClosestElementAncestor from '../utils/findClosestElementAncestor';
+import getTagOfNode from '../utils/getTagOfNode';
+import toArray from '../jsUtils/toArray';
+import VTable from '../table/VTable';
+import { NodePosition, NodeType, TableOperation } from 'roosterjs-editor-types';
+
+/**
+ * @internal
+ * @param root The table to paste into
+ * @param nodeToInsert A Node containing the table to be inserted
+ * @param position The position to paste the table
+ */
+export default function pasteTable(
+    root: HTMLElement,
+    nodeToInsert: Node,
+    position: NodePosition,
+    range: Range
+) {
+    if (
+        (nodeToInsert.childNodes.length == 1 &&
+            getTagOfNode(nodeToInsert.childNodes[0]) == 'TABLE') ||
+        (nodeToInsert.childNodes.length == 2 &&
+            getTagOfNode(nodeToInsert.childNodes[0]) == 'TABLE' &&
+            getTagOfNode(nodeToInsert.childNodes[1]) == 'BR')
+    ) {
+        let rootNodeToInsert: Node | null = nodeToInsert;
+
+        if (rootNodeToInsert.nodeType == NodeType.DocumentFragment) {
+            let rootNodes = toArray(rootNodeToInsert.childNodes).filter(
+                (n: ChildNode) => getTagOfNode(n) != 'BR'
+            );
+            rootNodeToInsert = rootNodes.length == 1 ? rootNodes[0] : null;
+        }
+
+        let tdNode = findClosestElementAncestor(position.node, root, 'TD,TH');
+
+        // This is the table on the clipboard
+        let newTable = new VTable(<HTMLTableElement>rootNodeToInsert);
+        // This table is already on the editor
+        let currentCell = new VTable(<HTMLTableCellElement>tdNode);
+
+        let currentRow = currentCell.row!;
+        let currentCol = currentCell.col!;
+
+        let rows = currentRow + newTable.cells?.length!;
+        let columns = currentCol + newTable.cells?.[0].length!;
+
+        while (currentCell.cells!.length! < rows) {
+            currentCell.edit(TableOperation.InsertBelow);
+        }
+
+        while (currentCell.cells![0].length < columns) {
+            currentCell.edit(TableOperation.InsertRight);
+        }
+
+        for (let i = currentRow; i < rows; i++) {
+            for (let j = currentCol; j < columns; j++) {
+                let cell = currentCell.getCell(i, j);
+                let newcell = newTable.getTd(i - currentRow, j - currentCol);
+                if (!cell.spanAbove && !cell.spanLeft) {
```

Comment:
Why we check spanAbove and spanLeft here? I think we can still do the replacement even if it is spanLeft or spanAbove?

Reply:
To skip the replacement as it can be more costly.

Reply:
How cost it is?

A simple loop here won't be a problem.

Reply:
And what if the source table cell has spanLeft/spanAbove but the target cell doesn't have?

Reply:
Based on Word online, the merged cells of the source table are not kept.

Reply:
Right, so we need to follow this behavior, and do not skip cells that has spanLeft/spanAbove since that pasted table cell may not be in that case

Reply:
I have to skip them if they have no content. I can just check for .td to exist instead of the spans, but I have to skip VCells that are empty.

Reply:
In those cases, source table cell won't have TD, but target table will have TD. If you skip the cells that don't have TD, the content from target table will be lost.

Since we are replacing everything from target table to source table, you don't need to check if a table cell is empty, but just move whatever we have from target table to source table. If it is empty, just move empty to source table.

Reply:
I may be confused on the terms of source and target, but it is fine if content you intend to paste gets lost due to merged cells.